### ISSUE TITLE
Upload the Android App to a F-Droid repository 

### ISSUE THREAD

**Issue by KoalaSat** [RoboSats/robosats#1261]
**Is your feature request related to a problem? Please describe.**
Users now have to access directly to the github page to be up to date with the Android app or using tools like obtanium. Would be great to have it available on a major F-Droid repository

**Comment by KoalaSat** [comment]
Hello @IzzySoft I was thinking on you while I was creating this issue, can we apply to have Robosats available in your repo? 😃 I'll take care of adding all the necessary configurations which I assume will be the same as we did for Nostros. Thanks in advance!

**Comment by IzzySoft** [comment]
I could include one of the per-ABI builds (arm64 probably). Scan results look good except for one thing that will at least need clarification:

```
Dangerous flags:
----------------
* usesCleartextTraffic
```

**Comment by KoalaSat** [comment]
> usesCleartextTraffic

Probably introduced by the Tor engine library, the app only works over the Tor network.

**Comment by KoalaSat** [comment]
> I could include one of the per-ABI builds (arm64 probably). Scan results look good except for one thing that will at least need clarification:
> 
> ```
> Dangerous flags:
> ----------------
> * usesCleartextTraffic
> ```

Probably introduced by the Tor engine library, the app only works over the Tor network.

**Comment by IzzySoft** [comment]
So it connects via a local (on-device) Tor proxy, with "cleartext" traffic just on-device?

**Comment by KoalaSat** [comment]
> So it connects via a local (on-device) Tor proxy, with "cleartext" traffic just on-device?

The proxy is loaded in-app with this library https://github.com/05nelsonm/kmp-tor/ and 100% of the traffic pass through it

**Comment by IzzySoft** [comment]
OK, that makes it easy:

```yaml
usesCleartextTraffic: used only on-device to communicate with the integrated Tor proxy
```

Guess you've nailed that even to `localhost` via network security policy? If not, you might want to do that (sorry, cannot help you with setting that up as that's beyond my knowledge).

That said: There's no APK at the latest release. Do you want me to pick up pre-releases then, or shall I wait for the next release having an APK?

**Comment by KoalaSat** [comment]
@IzzySoft I'll take a look and do some test thanks!

We are planning to create the release this weekend, also we have a PR with the fastlane metadata, we'll ping you once we are ready 😁 

Thank you again!

**Comment by KoalaSat** [comment]
Fastlane files ready for review https://github.com/RoboSats/robosats/pull/1270

**Comment by IzzySoft** [comment]
Reviewed :smiley: Just the `full_description.txt` needs a little formatting updates, as indicated.

**Comment by KoalaSat** [comment]
@IzzySoft we are ready to go 😁 https://github.com/RoboSats/robosats/releases/tag/v0.6.1-alpha

**Comment by IzzySoft** [comment]
Preparing the listing here then :smiley: Just one question left: Can the network service used be self-hosted? Is the server component FOSS? We might need the `NonFreeNet` or `TetheredNet` anti-feature here.

**Comment by IzzySoft** [comment]
PS: If you plan to get your app listed at F-Droid.org, the `fastlane` directory will need to go to the top level (root) of the repo. No issue for IzzyOnDroid where any custom location can be configured.

**Comment by KoalaSat** [comment]
> PS: If you plan to get your app listed at F-Droid.org, the `fastlane` directory will need to go to the top level (root) of the Okay @IzzySoft thanks for the hint we'll move it right now! 🚀

**Comment by IzzySoft** [comment]
Please let me know when done, as I then need to update your app's config here as well :wink: And do not forget to answer the other questions please (server part):

> Can the network service used be self-hosted? Is the server component FOSS? We might need the `NonFreeNet` or `TetheredNet` anti-feature here.

I know there are instructions for self-hosting – what is not clear is if one can choose that self-hosted server in the app itself. If not, it's `TetheredNet`. If the server part were not FOSS, it would be `NonFreeNet`.

Currently, your listing would look like this when "going live" around 6 pm UTC:

![image](https://github.com/RoboSats/robosats/assets/6781438/44922dc0-2ba5-49d9-951b-099fde510209)

(just the TetheredNet description was updated to "Server URL is currently hardcoded")

**Comment by KoalaSat** [comment]
> Can the network service used be self-hosted? 

Not from the android app, all frontend clients are hardcoded to use very specific URL and there is no way for an end user to change it. The self-hosted options are meant to always refer to the web client, not the backend.

> Is the server component FOSS? 

Asking @Reckless-Satoshi to confirm, but I would say yes! You can find the list of main services for backend here https://github.com/RoboSats/robosats/blob/main/docker-compose.yml

**Comment by IzzySoft** [comment]
Thanks, then I got that correct with `TetheredNet` – and all is ready for the go-live later today. After checking, you are welcome to [pick a badge](https://gitlab.com/IzzyOnDroid/repo/#if-my-app-is-available-here-do-you-have-a-badge-i-can-use-to-link-to-it) to link there e.g. from your Readme (check the details; guess you might want one of the shields to accompany those already present there; shields are available in the detailed instructions [here](https://gitlab.com/IzzyOnDroid/repo/-/wikis/API#shield); you might wish to increase the `cacheSeconds` to something like `3600` or higher as updates won't happen more often than once a day, so 1h caching should be fine).

**Comment by KoalaSat** [comment]
thank you so much for the support!! I'll ping you once we move `fastlane` to the root 😃

**Comment by Reckless-Satoshi** [comment]
> Can the network service used be self-hosted? Is the server component FOSS? We might need the `NonFreeNet` or `TetheredNet` anti-feature here.

Yes. RoboSats has 2 components, the client (for example the Android app is a full client) and the coordinator (we could say this is "the server"). There is many clients and many coordinators. The coordinators are, of course, fully FOSS (this monorepo contains the coordinator codebase as well).

Anyone can run a coordinator, the only current limitation is that coordinators are whitelisted and hardcoded into the client (new coordinators must open an Issue and a PR in this repo for inclusion in the client). 

In the near future, it would be trivial to add a feature for users to add new coordinators to their client through a form: although that might likely put users at risk...

**Comment by KoalaSat** [comment]
@IzzySoft `fastlane` has been move to root

**Comment by IzzySoft** [comment]
Eek, and forgive the blind me: `fastlane/metadata/android/<locale>` is where F-Droid.org would expect it :see_no_evil: See the freshly updated [IzzyOnDroid Fastlane Documentation](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Fastlane) for details…

**Comment by KoalaSat** [comment]
> Eek, and forgive the blind me: `fastlane/metadata/android/<locale>` is where F-Droid.org would expect it 🙈 See the freshly updated [IzzyOnDroid Fastlane Documentation](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Fastlane) for details…

Oops! Okay then I'll chsbge ir, is it required also forma Izzydroid? Should de do it asap?

**Comment by IzzySoft** [comment]
> is it required also for Izzydroid?

Nope, not at all (see the link in my previous comment: at IzzyOnDroid, *any* entry-point works as it's configured for each app explicitly. I've already adjusted it to the current location. It will just be needed if you decide for a listing at F-Droid.org.

And if you wonder, here are some details on that "entry-point":

* `fastlane/`: all the fastlane stuff goes here. Including the fastlane binaries, should you want to use them.
* `fastlane/metadata/`: here go the metadata. But there might be different platforms requiring different metadata, think e.g. of screenshots, so:
* `fastlane/metadata/android` / `fastlane/metadata/ios/`: the metadata for the specific platforms.

That's the "official" locations by Fastlane. F-Droid.org and IzzyOnDroid have some freedom of interpretation, so F-Droid.org e.g. also accepts `/metadata/<locale>` and IzzyOnDroid quite any place with the locales as next directory level.

**TL;DR:** Nothing urgent. Just let me know if and when you change it, and ideally before you make another release after having changed it (so the updater finds it).

**Comment by KoalaSat** [comment]
Great explanation! 👏👏

**Comment by IzzySoft** [comment]
Not before there's a new release including that :rofl: But yes. Please give me another ping then.

**Comment by KoalaSat** [comment]
> Not before there's a new release including that 🤣 But yes. Please give me another ping then.

We are good to go :rocket: https://github.com/RoboSats/robosats/releases/tag/v0.6.3-alpha

**Comment by IzzySoft** [comment]
Heh – the IoD updater pulled that already in yesterday, so it's even live already :partying_face:

**Comment by KoalaSat** [comment]
Thank you @IzzySoft !!! closing this issue :smiley:

**Comment by KoalaSat** [comment]
@IzzySoft I see the anti-feature is still appearing, any insight? Thanks!

**Comment by IzzySoft** [comment]
Oops – missed that "coordinator" = "server", and thus no longer tethered. Too many issues to follow in too many projects :see_no_evil: I just removed it, should be gone with the next sync around 6 pm UTC. Apologies for the delay!

**Comment by KoalaSat** [comment]
haha thank you for checking and for your work!!

**Comment by licaon-kter** [comment]
@KoalaSat got a recipe ready for F-Droid
```
  - versionName: 0.7.1-alpha
    versionCode: 5
    commit: c11c264654195c0d4b72eac5cf2ef1a133d3b6d3
    subdir: mobile/android
    sudo:
      - sysctl fs.inotify.max_user_watches=524288 || true
      - echo "deb https://deb.debian.org/debian bullseye main" > /etc/apt/sources.list.d/bullseye.list
      - apt-get update
      - apt-get install -y -t bullseye openjdk-11-jdk-headless
      - update-java-alternatives --set java-1.11.0-openjdk-amd64
      - apt-get install -y npm build-essential
    gradle:
      - yes
    output: app/build/outputs/apk/release/app-universal-release-unsigned.apk
    srclibs:
      - robo-identities@c95a9f563bf50377f82665661152ee42097e773c
      - rustup@1.27.1
    rm:
      - desktopApp
      - docker
    prebuild:
      - cd ../..
      - npm install --build-from-source
      - cd frontend
      - npm install --build-from-source
      - npm run build
      - cd ../mobile
      - npm install --build-from-source
      - cp -r patch_modules/* node_modules/
      - sdkmanager "build-tools;30.0.3" "platforms;android-33"
      - $$rustup$$/rustup-init.sh -y --default-toolchain 1.82.0 --target aarch64-linux-android
        armv7-linux-androideabi x86_64-linux-android
      - source $HOME/.cargo/env
    scanignore:
      - mobile/android/build.gradle
      - mobile/node_modules/jsc-android
      - mobile/node_modules/react-native/build.gradle.kts
      - mobile/node_modules/@react-native-community/netinfo/android/build.gradle
      - mobile/node_modules/react-native-encrypted-storage/android/build.gradle
      - mobile/node_modules/react-native-webview/android/build.gradle
      - mobile/patch_modules/react-native-encrypted-storage/android/build.gradle
    scandelete:
      - frontend/node_modules
      - mobile/android/app/src/main/jniLibs
      - mobile/node_modules
    build:
      - export ANDROID_NDK_ROOT=$$NDK$$
      - source $HOME/.cargo/env
      - cargo install cargo-ndk
      - pushd ../../frontend
      - npm run build
      - popd
      - pushd $$robo-identities$$
      - cd robonames
      - cargo ndk -t armeabi-v7a -t arm64-v8a -t x86_64 -o ./target build --release
      - cd ../robohash
      - cargo ndk -t armeabi-v7a -t arm64-v8a -t x86_64 -o ./target build --release
      - popd
      - mv $$robo-identities$$/target/armv7-linux-androideabi/release/librobo*.so
        app/src/main/jniLibs/armeabi-v7a/
      - mv $$robo-identities$$/target/x86_64-linux-android/release/librobo*.so app/src/main/jniLibs/x86_64/
      - mv $$robo-identities$$/target/aarch64-linux-android/release/librobo*.so app/src/main/jniLibs/arm64-v8a/
    ndk: 25.1.8937393
```

some questions:
* does it look ok to you in general? 
* do I need to run npm install in all those folders?
* which commit did you build robo-identities from? maybe there's a chance for reproducibility
* you have a libs/x86 folder but you don't build robo-identities for it? won't that crash on x86 devices?

**Comment by KoalaSat** [comment]
Ey @licaon-kter this is awesome thank you!

- The workflow looks great!
- Only for `/frontend` and `/mobile`
- This is the code https://github.com/RoboSats/robo-identities, in local it just works with `docker compose up`, so everything you need to launch it should be on the `docker-compose.yml`
- I can´t remember if we ever test it out :thinking: but what do you meant we don't build robo-identities for it?

**Comment by licaon-kter** [comment]
> This is the code https://github.com/RoboSats/robo-identities, in local it just works with docker compose up, so everything you need to launch it should be on the docker-compose.yml

we use our clean Debian VM, so no Docker, but we can mirror your Docker commands, the only thing that we'd like to match is exact commit the libs were build from: https://github.com/RoboSats/robo-identities/commits/main/ like see srclibs section, I used latest https://github.com/RoboSats/robo-identities/commit/c95a9f563bf50377f82665661152ee42097e773c but the .so differences are massive :disappointed: 

rust built libs can be made reproducible, afaik, so this might be possible, here's the full difflog: 
[robo5_final2.log.7z.zip](https://github.com/user-attachments/files/17621017/robo5_final2.log.7z.zip)


> I can´t remember if we ever test it out 🤔 but what do you meant we don't build robo-identities for it?

open the APK, look in libs/x86 folder, see no librobo*.so files and see no rust tooling, nor build command for x86: https://github.com/RoboSats/robo-identities/commit/c95a9f563bf50377f82665661152ee42097e773c#diff-e45e45baeda1c1e73482975a664062aa56f20c03dd9d64a827aba57775bed0d3R18 just x86_64

_(yes, Android x86 is rather rare these days but you added the rest of the libs so...)_

**Comment by KoalaSat** [comment]
I'm pretty sure this was the commit, before x86: https://github.com/RoboSats/robo-identities/commit/424a297c30beb71ad1c85a2a15b18d4cbfc92d55

 We can even update those files on Robosats with the output of the last commit to keep everything consistent

**Comment by KoalaSat** [comment]
Yes, we are planning to release another version soon, I can just rebuild those libs with the latest commits and take a look to that x86

**Comment by licaon-kter** [comment]
@KoalaSat ping when ready to test

**Comment by KoalaSat** [comment]
https://github.com/RoboSats/robosats/pull/1600

**Comment by licaon-kter** [comment]
https://github.com/m2049r/xmrwallet/releases/tag/v4.1.5 -> commit `989d52b33de200253f1ea3bc2d7d91b8764079fd`

download APK, unpack, `grep revision META-INF/version-control-info.textproto` -> `revision: "b08cd2fd2f18af55b0f1ee5003a88d8e81712e06"`

https://github.com/m2049r/xmrwallet/commit/b08cd2fd2f18af55b0f1ee5003a88d8e81712e06 -> 404

Wat?

**Comment by licaon-kter** [comment]
For a final checkup, final-ish recipe:
```
Binaries: https://github.com/RoboSats/robosats/releases/download/v%v/robosats-v0.7.3.alpha-universal.apk

Builds:
  - versionName: 0.7.3-alpha
    versionCode: 7
    commit: a600e70be019bdaade8594dbe9b4638b8a977959
    subdir: mobile/android
    sudo:
      - chown -R vagrant /root
      - sysctl fs.inotify.max_user_watches=524288 || true
      - echo "deb https://deb.debian.org/debian bullseye main" > /etc/apt/sources.list.d/bullseye.list
      - apt-get update
      - apt-get install -y -t bullseye openjdk-11-jdk-headless
      - update-java-alternatives --set java-1.11.0-openjdk-amd64
      - apt-get install -y build-essential npm
    gradle:
      - yes
    output: app/build/outputs/apk/release/app-universal-release-unsigned.apk
    srclibs:
      - robo-identities@c95a9f563bf50377f82665661152ee42097e773c
      - rustup@1.27.1
    rm:
      - desktopApp
      - docker
    prebuild:
      - export ndkversion=$(grep ndkVersion build.gradle | head -n1 | cut -d "\"" -f2)
      - '[[ $ndkversion ]] || exit 1'
      - sdkmanager "ndk;$ndkversion"
      - cd ../../frontend
      - npm install --build-from-source
      - cd ../mobile
      - npm install --build-from-source
      - cp -r patch_modules/* node_modules/
      - export CARGO_HOME=/root/.cargo
      - $$rustup$$/rustup-init.sh -y --default-toolchain nightly-2024-07-08 --target
        aarch64-linux-android armv7-linux-androideabi x86_64-linux-android
    scanignore:
      - mobile/android/build.gradle
      - mobile/node_modules/jsc-android
      - mobile/node_modules/react-native/build.gradle.kts
      - mobile/node_modules/@react-native-community/netinfo/android/build.gradle
      - mobile/node_modules/react-native-encrypted-storage/android/build.gradle
      - mobile/node_modules/react-native-webview/android/build.gradle
      - mobile/patch_modules/react-native-encrypted-storage/android/build.gradle
    scandelete:
      - frontend/node_modules
      - mobile/android/app/src/main/jniLibs
      - mobile/node_modules
    build:
      - export ndkversion=$(grep ndkVersion build.gradle | head -n1 | cut -d "\"" -f2)
      - export ANDROID_NDK_ROOT=/opt/android-sdk/ndk/$ndkversion
      - export ANDROID_NDK_HOME=$ANDROID_NDK_ROOT
      - pushd ../../frontend
      - npm run build
      - popd
      - export CARGO_HOME=/root/.cargo
      - source $CARGO_HOME/env
      - cargo install cargo-ndk
      - mv $$robo-identities$$/* /root/
      - pushd /root/robonames
      - cargo ndk -t armeabi-v7a -t arm64-v8a -t x86_64 -o ./target build --release
      - cd ../robohash
      - cargo ndk -t armeabi-v7a -t arm64-v8a -t x86_64 -o ./target build --release
      - popd
      - mv /root/target/armv7-linux-androideabi/release/librobo*.so app/src/main/jniLibs/armeabi-v7a/
      - mv /root/target/x86_64-linux-android/release/librobo*.so app/src/main/jniLibs/x86_64/
      - mv /root/target/aarch64-linux-android/release/librobo*.so app/src/main/jniLibs/arm64-v8a/

AllowedAPKSigningKeys: 21a2cef69f8a304945d1d57ab16747971c08d288c13e0756f0f65d6e70c9c924
```

from top to bottom some issues:
* Binaries is templated from versionCode, versionName, or a combo of these, your APK name in 0.7.3 is neither, but Izzy already picked this version up so you can't change the name, maybe for the next release?
* robo-identities commit used, can this be written somewhere in THIS repo?
* rust-default toolchain version, same as above

**Comment by KoalaSat** [comment]
> Binaries is templated from versionCode, versionName, or a combo of these, your APK name in 0.7.3 is neither, but Izzy already picked this version up so you can't change the name, maybe for the next release?

Alright, let's remember this @Reckless-Satoshi 

> robo-identities commit used, can this be written somewhere in THIS repo?
>rust-default toolchain version, same as above

What do you think about including it as part of the file name?

**Comment by licaon-kter** [comment]
we need to be able to grep the name from the source code tree of that release, not sure which file you have in mind

eg. when looking at https://github.com/RoboSats/robosats/tree/v0.7.3-alpha the hash would be there somewhere in a file

**Comment by IzzySoft** [comment]
```
- chown -R vagrant /root
- …
- mv /root/target/armv7-linux-androideabi/release/librobo*.so app/src/main/jniLibs/armeabi-v7a/
```

Are you building the app as root, @KoalaSat ? 😱

**Comment by KoalaSat** [comment]
> ```
> - chown -R vagrant /root
> - …
> - mv /root/target/armv7-linux-androideabi/release/librobo*.so app/src/main/jniLibs/armeabi-v7a/
> ```
> 
> Are you building the app as root, [@KoalaSat](https://github.com/KoalaSat) ? 😱

Hum, It shouldn´t, what this command does is to move the `.so` libraries from the `/root` folder to its place on the android folder. 

Am I missing something?

**Comment by licaon-kter** [comment]
I think the question is based on thu fact that you usually need to be user root to write stuff in folder /root

**Comment by IzzySoft** [comment]
What @licaon-kter wrote 😉 (thanks LK!). How do `.so` files end up in `/root` with your build, @KoalaSat?

```yaml
      - pushd /root/robonames
      - cargo ndk -t armeabi-v7a -t arm64-v8a -t x86_64 -o ./target build --release
      - cd ../robohash
      - cargo ndk -t armeabi-v7a -t arm64-v8a -t x86_64 -o ./target build --release
      - popd
```

That looks as if (on your end), cargo was installed as `root`, and those actions on your end are performed with the root user. For a successful RB, paths must be matched here, which is why LK set up the recipe that way (you can see further up that, in his recipe, LK made the `/root` directory owned by the building user, here `vagrant`, to avoid building as root – as especially the latter would be a very bad idea).

IMHO the `/root` directory should have nothing to do at all with building apps, this could even be a security issue: in an ephemeral container like licaon's (or on our builders at IzzyOnDroid), everything placed there is gone again after the build finishes. But if you do that on your machine, stuff would be placed in the realms of root, with root permissions. Stuff that _could_ include something malicious – and as a result compromise your whole machine. I'd strongly discourage from that.

**Comment by licaon-kter** [comment]
Also FYI, we need to recompile some wasm files, not sure what's the status there, ping @linsui

**Comment by linsui** [comment]
IIRC we don't need those wasm.

**Comment by KoalaSat** [comment]
Friendly ping for this issue, my understanding is that we are good so far are not included or used. 

I included commit hash to the apk name (ie: https://github.com/RoboSats/robosats/releases/tag/v0.8.1-alpha) and rebuild the versioning strategy https://github.com/RoboSats/robosats/issues/2147
